File: plugins/javascript/writer_test.js
var Writer = require("./dist/include/javascript/topcoder/writer");

exports.testWriteAll = function(test) { try {
    var writer = new Writer();
    writer.next("Seri", "String");
    writer.next();
    writer.next('M', "char");
    writer.next();
    writer.next(false, "boolean");
    writer.next();
    writer.next(20.14, "double");
    writer.next();
    writer.next([1, 2, 3, 4], "int[]");
    var result = "\"Seri\", 'M', false, 20.14, [1, 2, 3, 4]";
    test.strictEqual(writer.toString(), result);
    test.done();
} catch (err) {
    console.log(err.toString());
    console.trace();
}};

exports.testModule = function(test) { try {
    var tc = require("./dist/include/javascript/topcoder");
    var reader = new tc.Reader("abc");
    var writer = new tc.Writer();
    var err = new tc.UnsupportedType();
    test.done();
} catch (err) {
    console.log(err.toString());
    console.trace();
}};

File: plugins/ruby/writer_test.rb
require "test/unit"

require "gettc/types"
include Gettc

require "topcoder/writer"
require "topcoder/reader"
include TopCoder

class WriterTest < Test::Unit::TestCase
  def check(value, type)
    assert_equal Reader.new(Writer.new.next(value, type).to_s).next(type), value
  end

  def test_write_single_value
    check 20.14, TDouble
    check '@', TChar
    check 'String white a "', TString
    check [1, 2, 3, 4], TArray.new(TInt)
    check [[20.14]], TArray.new(TArray.new(TDouble))
    check [[[]]], TArray.new(TArray.new(TArray.new(TBoolean)))
    check false, TBoolean
    check [["Jon Snow"],['The "Little Finger"']], TArray.new(TArray.new(TString))

    assert_raise UnsupportedType do
      Writer.new.next(Hash.new, "dict")
    end
  end

  def test_write_all
    writer = Writer.new
    writer.next("Seri", TString)
    writer.next
    writer.next(20.14, TFloat)
    writer.next
    writer.next([1, 2, 3, 4], TArray.new(TInt))
    assert_equal '"Seri", 20.14, [1, 2, 3, 4]', writer.to_s
  end
end
